[빙봉] 로또 2단계 - 수동 구매 미션 제출합니다.

aegis1920 opened the pull request:
많은 피드백 부탁드립니다 :) 감사합니다!

giantim132 committed:
feat : readme 업데이트

giantim132 committed:
feat : LottoTicket 클래스 추가 / LottoTicket 생성자 테스트 코드 추가

giantim132 committed:
feat : WinningLottoTicket 클래스 생성 및 테스트 코드 작성

giantim132 committed:
feat : BonusBall 클래스 추가 및 생성자 테스트 코드 작성 / LottoNumbers 클래스 추가

giantim132 committed:
feat : LottoTicket 클래스 Collections.sort 추가

giantim132 committed:
feat : 로또 구매 개수 만큼 로또 티켓을 생성하는 메소드 구현 및 테스트

giantim132 committed:
feat : Money 클래스 생성 및 테스트 코드 작성

- 로또 티켓 갯수를 구해주는 calculateLottoTicket() 메서드 추가
- refactor : WinningLottoTicket 클래스에서 정규표현식 제거

aegis1920 committed:
feat : LottoTickets 클래스 생성 및 LottoTicket 클래스에 당첨 복권과 매칭해 카운트해주는 메서드 추가

- WinningLottoTicket 클래스에 private 접근제한자 추가

aegis1920 committed:
feat : LottoResult, LottoResults 클래스, LottoType Enum 및 테스트 코드 추가

aegis1920 committed:
feat : LottoCalculator 클래스 추가 및 테스트 코드 작성

aegis1920 committed:
feat : InputView 클래스 추가

aegis1920 committed:
feat : OutputView 클래스 추가 및 LottoCalculator static으로 잠시 변경

aegis1920 committed:
feat : LottoApplication 클래스 생성 및 메서드 수정

 - LottoNubmers 클래스명을 LottoNumbersGenerator로 변경
 - getTotal 메서드명 수정
 - getCorrectCount 메서드에서 인덱스로 비교하는 것을 contains로 수정

aegis1920 committed:
refactor : LottoResultsTest 테스트 코드 수정

aegis1920 committed:
refactor : 상수화 리펙토링 및 공백 제거, 쓸모없는 메서드 삭제

aegis1920 committed:
refactor : HashMap 변수명 수정, 테스트코드 수정, fix : isMatchBonusBall 메서드

 - String.Utils의 isBlank 메서드를 spark 패키지에서 apache 패키지로 변경

aegis1920 committed:
refactor : LottoApplication 메서드 분리, 공백 제거, 접근제한자 수정, LottoTicketsTest 클래스의 테스트 코드 수정

aegis1920 committed:
refactor : 숫자에 언더스코어 추가 및 LottoType 네이밍을 RankType으로 변경

aegis1920 committed:
refactor : validateNull 메서드 삭제, validateEmpty 메서드명을 validateBlank로 변경, WinningLottoTicket의 생성자에서 bonusBall을 받도록 수정

aegis1920 committed:
refactor : LottoApplication의 로직 수정, map의 선언부를 HashMap에서 Map으로 수정, LottoManager 클래스 추가, 테스트코드 수정

- LottoResult 클래스 명을 LottoCount로 변경
- LottoResults 클래스에서 add() 메서드 삭제
- LottoTicket 클래스에 있던 getCorrectCount() 메서드를 WinningLottoTicket 클래스로 옮김
- LottoTickets 클래스에 있던 match() 메서드 삭제
- LottoResultsTest 클래스 삭제
- LottoTicketsTest 클래스 삭제

aegis1920 committed:
refactor : LottoCalculator 클래스 및 테스트 코드 삭제, Money 클래스에 돈 관련 메서드 추가 및 테스트 코드 작성

aegis1920 committed:
refactor : BonusBall 클래스 생성자 수정, 보너스 볼 중복 체크 메서드를 WinningLottoTIcket으로 위치 변경

aegis1920 committed:
refactor : LottoCount 클래스 삭제, RankType enum에 RankType 반환 메서드 추가

- LottoCount에 대한 메서드 삭제
- LottoResults 클래스의 putLottoResults 메서드 수정
- OutputView에 printWinningStatistics 메서드에서 Map을 반복할 때 entry 방법이 아닌 foreach로 변경

aegis1920 committed:
refactor : addTicketNumber() 메서드명을 createWinningTicket() 으로 수정

- List를 메서드 안에서 만들고 반환하도록 로직 수정

aegis1920 committed:
feat : LottoNumber 클래스 추가 및 테스트코드 작성, 수정, refactor : LottoNumberGenerator 에서 같은 인스턴스로 가져오도록 getter 메서드 추가, OutputView 에서 Map이 아닌 RankType으로 출력 수정

aegis1920 committed:
refactor : 공백 컨벤션 지키기

aegis1920 committed:
refactor : LottoManager 클래스의 match() 메서드 수정, LottoResults 클래스 생성자 로직 변경 및 메서드 추가, LottoResults 테스트 코드 작성

aegis1920 committed:
refactor : LottoNumber 클래스에서 LottoNumber를 가져오기 위한 Map 자료형의 숫자 풀 추가, 나머지 클래스에서 getLottoNumber 메서드 수정, getLottoNumber 메서드 테스트 코드 추가

aegis1920 committed:
refactor : LottoApplication 클래스의 run() 메서드 로직 분리, LottoNumber 클래스의 클래스 변수명 수정 및 삭제, Money 클래스의 메서드 명 수정,

aegis1920 committed:
refactor : Money 클래스의 변수 네이밍 수정 및 for문 로직 수정

aegis1920 committed:
refactor : BonusBall 클래스 및 테스트 코드 삭제, BonusBall에 있던 로직을 LottoNumber에 추가, bonusBall 의 네이밍을 bonusNumber로 수정,

aegis1920 committed:
refactor : 테스트 코드의 DisplayName 어노테이션 수정

aegis1920 committed:
fix : merge conflicts

aegis1920 committed:
docs : README.md에 수동 구매 추가

aegis1920 committed:
feat : ManualBuyLottoTicketCount 클래스 및 테스트 코드 추가

aegis1920 committed:
feat : LottoTicket 클래스에 containsLottoNumber 메서드 추가 및 테스트 코드 작성, LottoTicketGenerator 클래스 추가 및 테스트코드 작성, refactor : LottoNumbersGenerator 클래스명 수정, WinningLottoTicket 의 생성자 로직 리펙토링

- LottoNumbersGenerator 클래스명을 LottoTicketsGenerator 로 변경
- LottoTicket 안에서 LottoNumber 를 찾을 수 있는 containsLottoNumber() 메서드 추가
- LottoTicketGenerator 클래스 추가 및 문자열로 받았을 때 로또 티켓을 생성할 수 있는 createLottoTicket() 메서드 추가
- WinningLottoTicket 클래스의 생성자 로직 리펙토링
-

aegis1920 committed:
feat : LottoTicketsGenerator 클래스에 addManualLottoTicket() 메서드 추가

aegis1920 committed:
refactor : LottoApplication 로직 수정 및 ManualBuyLottoTicketCount 클래스 명 변경 및 테스트 코드 추가, Money 클래스에 있는 countLottoTicket() 메서드 삭제, InputView 클래스에 수동 로또 입력 로직 작성, OutputView 클래스에 수동 로또 출력 로직 작성, MoneyTest 클래스에서 구매한 로또 총 갯수를 테스트하는 로직 삭제

- ManualBuyLottoTicketCount 클래스를 ManualLottoTicketQuantity 로 변경

aegis1920 committed:
refactor : DisplayName 어노테이션 Should, When 규칙에 따라 작성

jihan805 commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -33,13 +72,17 @@ public boolean equals(Object o) {
-    @Override
-    public int hashCode() {
-        return Objects.hash(lottoNumber);
+    public static LottoNumber getLottoNumber(int number) {
```
get~하는대신, LottoNumber 생성을 위한 정적 팩터리 메서드를 만들어 보면 어떨까요?

jihan805 commented on src/main/java/domain/LottoApplication.java:
```diff
@@ -6,14 +6,29 @@
 
-    public static LottoTickets generateLottoTicketsByMoney(Money money) {
-        return new LottoTickets(LottoNumbersGenerator.generateLottoTickets(money.calculateLottoTicket()));
+    public static void LottoTicketsResult(Money money, LottoTickets lottoTickets) {
```
메서드 네이밍 lottoTicketsResult~, 아래 모두 동일!

jihan805 commented on src/main/java/domain/LottoApplication.java:
```diff
@@ -6,14 +6,29 @@
+        OutputView.printWinningStatistics(LottoManager.match(lottoTickets, inputWinningLottoTicket()), money);
+    }
+
+    public static void LottoTicketsCount(Money money, ManualLottoTicketQuantity manualLottoTicketQuantity) {
```
메서드로 분리하고 싶다면, buyManualLotto등의 네이밍이 더 적합하지 않을까요?

jihan805 commented on src/main/java/domain/LottoApplication.java:
```diff
@@ -6,14 +6,29 @@
+        OutputView.printBuyTicketCount(money, manualLottoTicketQuantity);
+    }
+
+    public static void LottoTicketsStatus(LottoTickets lottoTickets) {
```
이와같은 경우, 메서드로 분리하지 않아도 괜찮습니다 :) 
이전 피드백의 경우 
OutputView.printWinningStatistics(LottoManager.match(lottoTickets,inputWinningLottoTicket()).getLottoResults(), money.getMoney())
이 한줄에 축약된 행위들이 너무 많아, 조금 분리하여 가독성을 높이면 어떨까?에 대한 피드백이었습니다. 

LottoTicketsResult() 메서드 역시 동일하구요 !!

jihan805 commented on src/main/java/view/InputView.java:
```diff
@@ -10,6 +13,18 @@ public static String inputBuyMoney() {
+    public static void inputManualLottoTicket(ManualLottoTicketQuantity manualLottoTicketQuantity) {
+        System.out.println("수동으로 구매할 번호를 입력해주세요.");
+        for (int i = 0; i < manualLottoTicketQuantity.getManualLottoTicketQuantity(); i++) {
+            LottoTicketsGenerator.addManualLottoTicket(scanner.nextLine());
```
수동으로 구매한 로또를 InputView에서 생성하는 것이 적절할까요?
InputView 는 입력받는 이상의 역할을 하지 않도록 해요!

jihan805 commented on src/main/java/domain/LottoTicketsGenerator.java:
```diff
@@ -6,30 +6,31 @@
 
-    private static final List<LottoNumber> originLottoNumbers;
     private static final List<LottoNumber> shuffleLottoNumbers;
+    private static final List<LottoTicket> lottoTickets = new ArrayList<>();
```
LottoTicketsGenerator가 lottoTickets을 가지고 있을 필요는 없을것 같아요!
아래 힌트를 참고해 수동 & 자동 로또 구매를 위한 설계를 조금 더 고민해보면 어떨까요?

2단계에서 3단계로 요구사항이 변경될 때 로또를 생성하는 부분의 요구사항만 변경되었어요,
로또를 생성하는 부분과 로또 결과를 구하는 부분을 다음과 같은 인터페이스로 분리해 구현해 보면 어떨까요?
(2단계에서 3단계로 요구사항이 변경될 때 변경을 최소화하는 연습)

```
public interface LottosGenerator {
    List<Lotto> generate();
}
```

jihan805 reviewed (changes requested):
안녕하세요 전 단계 피드백 반영 잘하셨어요 💯 
몇가지 피드백 남겼으니 확인 해주세요 :)

aegis1920 committed:
refactor : LottoNumber 클래스에 정적 팩토리 메서드 추가, from 메서드 추가, validate 메서드 수정, WinningLottoTicket 클래스에 중복되는 validate 메서드 삭제

aegis1920 committed:
refactor : LottoApplication 클래스의 메서드 명 수정

aegis1920 committed:
feat : AutoCount, ManualCount 클래스 추가, AutoLottoGenerator, ManualLottoGenerator 클래스 추가, refactor : InputView 에 비즈니스 로직이 들어가지 않도록 수정, LottoManager 에서 generate 하도록 수정

- AutoCount, ManualCount 클래스를 추가해 LottoTicketCount 의 상태값들을 포장
- AutoLottoGenerator, ManualLottoGenerator 클래스를 추가하고 LottoTicketsGenerator 를 인터페이스로 만들어 구현하도록 함
- InputView와 OutputView 리펙토링

aegis1920 committed:
refactor : 상수 이름 수정

aegis1920 committed:
refactor : LottoNumber 클래스의 equals 컨벤션, LottoTickets 의 containsLottoNumber 메서드 명 수정, Money 클래스의 getTotalWinningPrice, getProfit 메서드 명 수정, WinningLottoTicket 클래스의 getCorrectCount 메서드 명 수정

aegis1920 committed:
refactor : Money 클래스, RankType Enum의 int 형을 long 형으로 바꿈

- 상금이 21억을 넘을 경우를 대비해 int 형을 long 형으로 바꿔줌

aegis1920 committed:
refactor : RankType에 있던 스트링과 expression 을 빼고 상금만 넣어 Enum에 데이터만 들어가도록 함

- OutputView, Money의 역할에 맞게 수정

jihan805 commented on src/main/java/domain/LottoNumber.java:
```diff
@@ -1,45 +1,89 @@
-    @Override
-    public int hashCode() {
-        return Objects.hash(lottoNumber);
+    public static LottoNumber from(String input) {
```
두 메서드 모두 생성자 아래로 위치

jihan805 commented on src/main/java/domain/LottoManager.java:
```diff
@@ -1,10 +1,27 @@
+        return new LottoResults(lottoTickets, winningLottoTicket);
+    }
+
+    public static LottoTickets generateLottoTickets(List<String> manualLottoTickets, AutoCount autoCount) {
```
인터페이스를 이용하여 구현 👍 
혹은 두 로또를 합쳐 하나의 로또를 만들어주는 클래스를 분리할수도 있을 것 같아요!

jihan805 commented on src/main/java/domain/ManualLottoGenerator.java:
```diff
@@ -0,0 +1,29 @@
+
+public class ManualLottoGenerator implements LottoTicketsGenerator {
+
+    List<String> manualLottoTickets;
```
특별한 경우가 아니라면 private 제어자를 사용하도록 해요

jihan805 commented on src/main/java/domain/RankType.java:
```diff
@@ -1,31 +1,28 @@
+    MATCH_FOUR(4, 50_000),
+    MATCH_FIVE(5, 1_500_000),
+    MATCH_FIVE_WITH_BONUS(5, 30_000_000),
+    MATCH_SIX(6, 2_000_000_000);
```
private boolean hasBonus를 가지고, 
MATCH_FIVE(5, 1_500_000,false),
MATCH_FIVE_WITH_BONUS(5, 30_000_000, true)
와 같이 구현하는 방법도 있을 것같아요!

jihan805 commented on src/test/java/domain/WinningLottoTicketTest.java:
```diff
@@ -36,27 +35,27 @@ void winningNumberConstructorTest(String input) {
     }
 
-    @DisplayName("당첨 복권 번호와 얼마나 맞는지 카운트해주는 메서드 테스트")
+    @DisplayName("Should_당첨 카운트(보너스 볼 포함)_When_로또 티켓과 당첨 복권이 주어졌을 때")
```
무엇을 테스트하여 기대하는지를 네이밍하는 것이 중요하지, Should와 When을 꼭 기입하지는 않아도되요! ㅎㅎ

jihan805 reviewed (approved):
피드백 반영 잘해주셨네요 👍 
머지할게요 :)
